pathchk/pathchk.1
.\"
.\" Copyright (c) 2005 Gunnar Ritter
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.\" Sccsid @(#)pathchk.1	1.1 (gritter) 7/17/05
.TH PATHCHK 1 "7/17/05" "Heirloom Toolchest" "User Commands"
.SH NAME
pathchk \- check pathnames
.SH SYNOPSIS
\fBpathchk\fR [\fB\-p\fR] \fIname\fR ...
.SH DESCRIPTION
The
.I pathchk
command verifies the validity or portability of path names.
For each
.I name
given, it is checked
that each of the intermediate directories is searchable
or could be created,
that none of its components is longer than NAME_MAX,
and that the whole
.I name
is not longer than PATH_MAX.
.PP
With the
.B \-p
option, the checks are performed using
_POSIX_NAME_MAX and _POSIX_PATH_MAX,
and each character of a component of
.I name
must be in the portable character set
(i.\|e. alphanumerics, dot `.', underscore `_', and minus sign `\-').
The check for local searchability is omitted.
.PP
If any violation of these rules is detected,
.I pathchk
prints an error message and exits with failure.


CHANGES
Release ...
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants. 

Release 230503 
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.

pathchk/pathchk.c
/*
 * pathchk - check pathnames
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, July 2005.
 */
/*
 * Copyright (c) 2005 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)pathchk.c	1.3 (gritter) 8/15/05";

#include <stdio.h>
#include <stdlib.h>
#include <libgen.h>
#include <limits.h>
#include <unistd.h>
#include <errno.h>
#include <asciitype.h>

#ifndef	_POSIX_NAME_MAX
#define	_POSIX_NAME_MAX	14
#endif	/* !_POSIX_NAME_MAX */
#ifndef	NAME_MAX
#define	NAME_MAX	255
#endif

static const char	*progname;		/* argv[0] to main() */
static int		status;			/* exit status */
static int		pflag;			/* -p option */
static long		path_max = PATH_MAX;
static long		name_max = NAME_MAX;

static void
usage(void)
{
	fprintf(stderr, "Usage: %s [-p] name ...\n", progname);
	exit(2);
}

static void
pathchk(char *name)
{
	char	*np = name, *nq = name;
	int	c, ac = 0;

	do {
		if (*np == '/') {
			if (pflag == 0) {
				*np = '\0';
				c = access(name, X_OK);
				*np = '/';
				if (c < 0 && errno != ENOENT && ac == 0) {
					fprintf(stderr, "%s: component "
							"\"%.*s\" of name "
							"\"%s\" is not "
							"searchable\n",
						progname, 
						np - name, name,
						name);
					status |= 1;
					ac = 1;
				}
			}
		}
		if (*np == '/' || *np == '\0') {
			if (np - nq >= name_max) {
				fprintf(stderr, "%s: component \"%.*s\" of "
						"name \"%s\" is longer than "
						"%ld bytes\n",
					progname, np - nq, nq, name, name_max);
				status |= 1;
			}
			nq = np;
		} else if (pflag) {
			if (!alnumchar(*np & 0377) && *np != '.' &&
					*np != '_' && *np != '-') {
				fprintf(stderr, "%s: character '\\%o' of name "
						"\"%s\" is not in the portable "
						"filename character set\n",
					progname, *np&0377, name);
				status |= 1;
			}
		}
	} while (*np++ != '\0');
	if (np - name >= path_max) {
		fprintf(stderr, "%s: name \"%s\" is longer than %ld bytes\n",
			progname, name, path_max);
		status |= 1;
	}
}

int
main(int argc, char **argv)
{
	int	c;

	progname = basename(argv[0]);
	while ((c = getopt(argc, argv, "p")) != EOF) {
		switch (c) {
		case 'p':
			pflag = 1;
			path_max = _POSIX_PATH_MAX;
			name_max = _POSIX_NAME_MAX;
			break;
		default:
			usage();
		}
	}
	if (optind == argc)
		usage();
	do
		pathchk(argv[optind]);
	while (++optind < argc);
	return status;
}
